File: backend/app/api/v1/organizations.py
import uuid

from fastapi import APIRouter

from app.core.authz import require_org_member, require_org_owner
from app.core.deps import CurrentUser, DbSession
from app.schemas.auth import MessageResponse
from app.schemas.organization import (
    MemberAddRequest,
    MemberResponse,
    OrganizationCreateRequest,
    OrganizationResponse,
    OrganizationUpdateRequest,
    PublicOrganizationResponse,
)
from app.services import organization as org_service

router = APIRouter(tags=["organizations"])


@router.post("/organizations", status_code=201)
async def create_organization(
    body: OrganizationCreateRequest, user: CurrentUser, db: DbSession
) -> OrganizationResponse:
    return await org_service.create_organization(db, user, body)


@router.get("/organizations/{slug}")
async def public_organization(slug: str, db: DbSession) -> PublicOrganizationResponse:
    return await org_service.get_public_organization(db, slug)


@router.get("/orgs/{org_id}")
async def get_organization(
    org_id: uuid.UUID, user: CurrentUser, db: DbSession
) -> OrganizationResponse:
    org = await require_org_member(db, org_id, user)
    return org_service.to_admin_response(org)


@router.put("/orgs/{org_id}")
async def update_organization(
    org_id: uuid.UUID, body: OrganizationUpdateRequest, user: CurrentUser, db: DbSession
) -> OrganizationResponse:
    org = await require_org_owner(db, org_id, user)
    return await org_service.update_organization(db, org, body)


@router.get("/orgs/{org_id}/members")
async def list_members(
    org_id: uuid.UUID, user: CurrentUser, db: DbSession
) -> list[MemberResponse]:
    await require_org_member(db, org_id, user)
    return await org_service.list_members(db, org_id)


@router.post("/orgs/{org_id}/members", status_code=201)
async def add_member(
    org_id: uuid.UUID, body: MemberAddRequest, user: CurrentUser, db: DbSession
) -> MemberResponse:
    await require_org_owner(db, org_id, user)
    return await org_service.add_member(db, org_id, body)


@router.delete("/orgs/{org_id}/members/{user_id}")
async def remove_member(
    org_id: uuid.UUID, user_id: uuid.UUID, user: CurrentUser, db: DbSession
) -> MessageResponse:
    await require_org_owner(db, org_id, user)
    await org_service.remove_member(db, org_id, user_id)
    return MessageResponse(message="除名しました")

File: frontend/src/api/organizations.ts
import { apiFetch } from "./client";
import type {
  Member,
  MessageResponse,
  Organization,
  PublicOrganization,
} from "./types";

export const createOrganization = (body: {
  name: string;
  slug: string;
  description?: string;
}) => apiFetch<Organization>("/organizations", { method: "POST", body });

export const getPublicOrganization = (slug: string) =>
  apiFetch<PublicOrganization>(`/organizations/${slug}`);

export const getOrganization = (orgId: string) =>
  apiFetch<Organization>(`/orgs/${orgId}`);

export const updateOrganization = (
  orgId: string,
  body: { name: string; description: string | null; website_url: string | null },
) => apiFetch<Organization>(`/orgs/${orgId}`, { method: "PUT", body });

export const listMembers = (orgId: string) =>
  apiFetch<Member[]>(`/orgs/${orgId}/members`);

export const addMember = (
  orgId: string,
  body: { email: string; role: "owner" | "member" },
) => apiFetch<Member>(`/orgs/${orgId}/members`, { method: "POST", body });

export const removeMember = (orgId: string, userId: string) =>
  apiFetch<MessageResponse>(`/orgs/${orgId}/members/${userId}`, {
    method: "DELETE",
  });
